### PR TITLE
[ADDED] Support for AES-GCM as a cipher along with ChaChaPoly. 

### PR DESCRIPTION
**Pull request by derekcollison**
Also supports upgrade to different cipher for existing store assets.

Signed-off-by: Derek Collison <derek@nats.io>

/cc @nats-io/core

### CODE REVIEW

File: server/filestore.go [hunk 1]
```diff
@@ -453,31 +474,71 @@ func (fs *fileStore) genEncryptionKeys(context string) (aek cipher.AEAD, bek *ch
 	if err != nil {
 		return nil, nil, nil, nil, err
 	}
-	kek, err := chacha20poly1305.NewX(rb)
+
+	sc := fs.fcfg.Cipher
+
+	var kek cipher.AEAD
+	if sc == ChaCha {
```
Comment (kozlovic) [inline server/filestore.go:481]: Looks like this block repeats 11 times in this PR, I think it justifies being made a function? Just pass the sc and rb.

**Review by kozlovic — CHANGES_REQUESTED** [review]

File: server/filestore.go (hunk 1 above)
Comment (derekcollison) [reply server/filestore.go:481]: Will take a look.

File: server/filestore.go [hunk 2]
```diff
@@ -453,31 +493,60 @@ func (fs *fileStore) genEncryptionKeys(context string) (aek cipher.AEAD, bek *ch
...
-	if n, err := rand.Read(seed); err != nil || n != 32 {
+
+	const seedSize = 32
+	seed = make([]byte, seedSize)
+	if n, err := rand.Read(seed); err != nil || n != seedSize {
 		return nil, nil, nil, nil, err
 	}
-	aek, err = chacha20poly1305.NewX(seed)
+
+	if sc == ChaCha {
```
Comment (kozlovic) [inline server/filestore.go:511]: Use the new function that you created?

File: server/filestore.go [hunk 3]
```diff
@@ -333,6 +354,10 @@ func newFileStoreWithCreated(fcfg FileStoreConfig, cfg StreamConfig, created tim
 			if err := fs.writeStreamMeta(); err != nil {
 				return nil, err
 			}
+		} else if err != nil {
+			return nil, err
+		} else {
+			// Check to make sure we can read our key file.
```
Comment (kozlovic) [inline server/filestore.go:360]: Something missing here or what?

File: server/filestore.go [hunk 4]
```diff
@@ -611,28 +681,44 @@ func (fs *fileStore) recoverMsgBlock(fi os.FileInfo, index uint32) (*msgBlock, e
...
-				return nil, err
+				// We may be here on a cipher conversion, so attempt to convert.
+				if err = mb.convertCipher(); err != nil {
+					return nil, err
+				}
+			} else {
+				mb.seed, mb.nonce = seed, ekey[:ns]
+			}
+
+			if sc == ChaCha {
```
Comment (kozlovic) [inline server/filestore.go:709]: Again, we can call the new function here I think.

File: server/filestore.go [hunk 5]
```diff
@@ -5571,21 +5727,40 @@ func (fs *fileStore) ConsumerStore(name string, cfg *ConsumerConfig) (ConsumerSt
...
+			nonce := ekey[:ns]
+			seed, err := kek.Open(nil, nonce, ekey[ns:], nil)
 			if err != nil {
-				return nil, err
+				// We may be here on a cipher conversion, so attempt to convert.
+				if err = o.convertCipher(); err != nil {
+					return nil, err
+				}
+			} else {
+				if sc == ChaCha {
```
Comment (kozlovic) [inline server/filestore.go:5753]: I don't see the use of the new function...

File: server/jetstream.go [hunk 6]
```diff
@@ -252,11 +250,20 @@ func (s *Server) decryptMeta(ekey, buf []byte, acc, context string) ([]byte, err
 	if err != nil {
 		return nil, err
 	}
-	aek, err := chacha20poly1305.NewX(seed[:])
+	var aek cipher.AEAD
+	if sc == ChaCha {
```
Comment (kozlovic) [inline server/jetstream.go:254]: Use the new function

File: server/filestore.go [hunk 7]
```diff
@@ -5640,6 +5815,74 @@ func (fs *fileStore) ConsumerStore(name string, cfg *ConsumerConfig) (ConsumerSt
...
+	}
+	ns := kek.NonceSize()
+	nonce := ekey[:ns]
+	seed, err := kek.Open(nil, nonce, ekey[ns:], nil)
+	if err != nil {
+		return err
+	}
+
+	var aek cipher.AEAD
+	if sc == AES {
```
Comment (kozlovic) [inline server/filestore.go:5853]: Can use the new function here too...

**Review by kozlovic — CHANGES_REQUESTED** [review]

File: server/filestore.go (hunk 2 above)
Comment (kozlovic) [reply server/filestore.go:507]: Still missing...

File: server/filestore.go (hunk 4 above)
Comment (kozlovic) [reply server/filestore.go:705]: That one too.

File: server/filestore.go (hunk 5 above)
Comment (kozlovic) [reply server/filestore.go:5749]: That one too.

File: server/filestore.go (hunk 7 above)
Comment (kozlovic) [reply server/filestore.go:5849]: Sorry, this is conversion so you are reversing things...

File: server/jetstream.go (hunk 6 above)
Comment (kozlovic) [reply server/jetstream.go:254]: Use new function.

**Review by kozlovic — CHANGES_REQUESTED** [review]
Looks like you still missed some places where you can use the function, however, the one with the conversion can't use it, but maybe adding a comment here to make it clear!

**Comment by derekcollison** [comment]
Can't use it for aek which requires a seed. kek requires rb and sc. I could add a convenience function for aeks if you feel strongly about it.

**Comment by kozlovic** [comment]
But you pass a byte slice:`rb` and `seed` are both `[]byte` so what's the difference? Meaning it looks like the same chacha or aes functions are invoked in both cases, just that the param passed to them is different byte slice content, but...

Here is a line where I commented that it is still missing: https://github.com/nats-io/nats-server/pull/3371/files#diff-384c189826934c9a6fc3554dafc63dab2076245010e3d6fce5c71a93e15e9877R507
```
if sc == ChaCha {
	aek, err = chacha20poly1305.NewX(seed)
} else if sc == AES {
	block, e := aes.NewCipher(seed)
	if e != nil {
		return nil, nil, nil, nil, err
	}
	aek, err = cipher.NewGCMWithNonceSize(block, block.BlockSize())
}
```
how is that code different from the function:
```
func genKEK(sc StoreCipher, rb []byte) (kek cipher.AEAD, err error) {
	if sc == ChaCha {
		kek, err = chacha20poly1305.NewX(rb)
	} else if sc == AES {
		block, e := aes.NewCipher(rb)
		if e != nil {
			return nil, err
		}
		kek, err = cipher.NewGCMWithNonceSize(block, block.BlockSize())
	} else {
		err = errUnknownCipher
	}
	return kek, err
}
```

**Comment by derekcollison** [comment]
Will take another look.

**Comment by derekcollison** [comment]
You are correct, updated. Thanks!

**Review by kozlovic — APPROVED** [review]
LGTM

**Pull request merged**